File: clients/javascript/src/vibe.ts
import { BiDiClient, BrowsingContextTree, NavigationResult, ScreenshotResult } from './bidi';
import { ClickerProcess } from './clicker';
import { Element, ElementInfo } from './element';
import { debug } from './utils/debug';

export interface FindOptions {
  /** Timeout in milliseconds to wait for element. Default: 30000 */
  timeout?: number;
}

export interface RecordingOptions {
  /** Frames per second. Default: 10 */
  fps?: number;
  /** Output format: 'mp4' or 'webm'. Default: 'mp4' */
  format?: 'mp4' | 'webm';
  /** Output file path. If not provided, uses temp directory */
  outputPath?: string;
}

interface StartRecordingResult {
  started: boolean;
  fps: number;
  format: string;
}

interface StopRecordingResult {
  stopped: boolean;
  outputPath: string;
}

interface VibiumFindResult {
  tag: string;
  text: string;
  box: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
}

export class Vibe {
  private client: BiDiClient;
  private process: ClickerProcess | null;
  private context: string | null = null;

  constructor(client: BiDiClient, process: ClickerProcess | null) {
    this.client = client;
    this.process = process;
  }

  private async getContext(): Promise<string> {
    if (this.context) {
      return this.context;
    }

    const tree = await this.client.send<BrowsingContextTree>('browsingContext.getTree', {});
    if (!tree.contexts || tree.contexts.length === 0) {
      throw new Error('No browsing context available');
    }

    this.context = tree.contexts[0].context;
    return this.context;
  }

  async go(url: string): Promise<void> {
    debug('navigating', { url });
    const context = await this.getContext();
    await this.client.send<NavigationResult>('browsingContext.navigate', {
      context,
      url,
      wait: 'complete',
    });
    debug('navigation complete', { url });
  }

  async screenshot(): Promise<Buffer> {
    const context = await this.getContext();
    const result = await this.client.send<ScreenshotResult>('browsingContext.captureScreenshot', {
      context,
    });
    return Buffer.from(result.data, 'base64');
  }

  /**
   * Execute JavaScript in the page context.
   */
  async evaluate<T = unknown>(script: string): Promise<T> {
    const context = await this.getContext();
    const result = await this.client.send<{
      type: string;
      result: { type: string; value?: T };
    }>('script.callFunction', {
      functionDeclaration: `() => { ${script} }`,
      target: { context },
      arguments: [],
      awaitPromise: true,
      resultOwnership: 'root',
    });

    return result.result.value as T;
  }

  /**
   * Find an element by CSS selector.
   * Waits for element to exist before returning.
   */
  async find(selector: string, options?: FindOptions): Promise<Element> {
    debug('finding element', { selector, timeout: options?.timeout });
    const context = await this.getContext();

    const result = await this.client.send<VibiumFindResult>('vibium:find', {
      context,
      selector,
      timeout: options?.timeout,
    });

    const info: ElementInfo = {
      tag: result.tag,
      text: result.text,
      box: result.box,
    };
    debug('element found', { selector, tag: result.tag });

    return new Element(this.client, context, selector, info);
  }

  /**
   * Start recording the browser session as a video.
   * Requires FFmpeg to be installed on the system.
   *
   * @param options - Recording options (fps, format, outputPath)
   * @returns Promise that resolves when recording starts
   *
   * @example
   * ```typescript
   * await vibe.startRecording({ fps: 10, format: 'mp4' });
   * // ... perform actions ...
   * const videoPath = await vibe.stopRecording();
   * ```
   */
  async startRecording(options?: RecordingOptions): Promise<void> {
    debug('starting recording', { fps: options?.fps, format: options?.format, outputPath: options?.outputPath });
    const context = await this.getContext();

    await this.client.send<StartRecordingResult>('vibium:startRecording', {
      context,
      fps: options?.fps,
      format: options?.format,
      outputPath: options?.outputPath,
    });

    debug('recording started');
  }

  /**
   * Stop recording and save the video file.
   *
   * @returns Promise that resolves with the path to the saved video file
   *
   * @example
   * ```typescript
   * const videoPath = await vibe.stopRecording();
   * console.log('Video saved to:', videoPath);
   * ```
   */
  async stopRecording(): Promise<string> {
    debug('stopping recording');
    const context = await this.getContext();

    const result = await this.client.send<StopRecordingResult>('vibium:stopRecording', {
      context,
    });

    debug('recording stopped', { outputPath: result.outputPath });
    return result.outputPath;
  }

  async quit(): Promise<void> {
    await this.client.close();
    if (this.process) {
      await this.process.stop();
    }
  }
}

File: clients/python/src/vibium/vibe.py
"""Vibe class - the main browser automation interface."""

import base64
from dataclasses import dataclass
from typing import Literal, Optional

from .client import BiDiClient
from .clicker import ClickerProcess
from .element import BoundingBox, Element, ElementInfo


@dataclass
class RecordingOptions:
    """Options for video recording."""

    fps: int = 10
    """Frames per second. Default: 10"""

    format: Literal["mp4", "webm"] = "mp4"
    """Output format. Default: 'mp4'"""

    output_path: Optional[str] = None
    """Output file path. If not provided, uses temp directory."""


class Vibe:
    """Main browser automation interface.

    Provides methods to navigate, interact with elements, and take screenshots.
    """

    def __init__(self, client: BiDiClient, process: Optional[ClickerProcess] = None):
        self._client = client
        self._process = process
        self._context: Optional[str] = None

    async def _get_context(self) -> str:
        """Get the browsing context ID."""
        if self._context:
            return self._context

        result = await self._client.send("browsingContext.getTree", {})
        contexts = result.get("contexts", [])
        if not contexts:
            raise RuntimeError("No browsing context available")

        self._context = contexts[0]["context"]
        return self._context

    async def go(self, url: str) -> None:
        """Navigate to a URL.

        Args:
            url: The URL to navigate to.
        """
        context = await self._get_context()
        await self._client.send(
            "browsingContext.navigate",
            {
                "context": context,
                "url": url,
                "wait": "complete",
            },
        )

    async def screenshot(self) -> bytes:
        """Capture a screenshot of the viewport.

        Returns:
            PNG image data as bytes.
        """
        context = await self._get_context()
        result = await self._client.send(
            "browsingContext.captureScreenshot",
            {"context": context},
        )
        return base64.b64decode(result["data"])

    async def find(self, selector: str, timeout: Optional[int] = None) -> Element:
        """Find an element by CSS selector.

        Waits for the element to exist before returning.

        Args:
            selector: CSS selector.
            timeout: Timeout in milliseconds (default: 30000).

        Returns:
            An Element instance.
        """
        context = await self._get_context()

        params = {
            "context": context,
            "selector": selector,
        }
        if timeout is not None:
            params["timeout"] = timeout

        result = await self._client.send("vibium:find", params)

        box_data = result["box"]
        info = ElementInfo(
            tag=result["tag"],
            text=result["text"],
            box=BoundingBox(
                x=box_data["x"],
                y=box_data["y"],
                width=box_data["width"],
                height=box_data["height"],
            ),
        )

        return Element(self._client, context, selector, info)

    async def start_recording(
        self,
        fps: int = 10,
        format: Literal["mp4", "webm"] = "mp4",
        output_path: Optional[str] = None,
    ) -> None:
        """Start recording the browser session as a video.

        Requires FFmpeg to be installed on the system.

        Args:
            fps: Frames per second. Default: 10
            format: Output format ('mp4' or 'webm'). Default: 'mp4'
            output_path: Output file path. If not provided, uses temp directory.

        Example:
            >>> await vibe.start_recording(fps=10, format='mp4')
            >>> # ... perform actions ...
            >>> video_path = await vibe.stop_recording()
        """
        context = await self._get_context()

        params = {
            "context": context,
            "fps": fps,
            "format": format,
        }
        if output_path is not None:
            params["outputPath"] = output_path

        await self._client.send("vibium:startRecording", params)

    async def stop_recording(self) -> str:
        """Stop recording and save the video file.

        Returns:
            Path to the saved video file.

        Example:
            >>> video_path = await vibe.stop_recording()
            >>> print(f"Video saved to: {video_path}")
        """
        context = await self._get_context()

        result = await self._client.send(
            "vibium:stopRecording",
            {"context": context},
        )
        return result["outputPath"]

    async def quit(self) -> None:
        """Close the browser and clean up resources."""
        await self._client.close()
        if self._process:
            await self._process.stop()
